pkg/pool-stable-phantom/test/StablePhantomPool.test.ts

```diff
@@ -759,6 +759,75 @@ describe('StablePhantomPool', () => {
         await expect(pool.exit({ data: wrongUserData })).to.be.revertedWith('Transaction reverted without a reason');
       });
 
+      describe('exit BPT in for one token out', () => {
+        let tokenIndex: number;
+        let token: Token;
+
+        sharedBeforeEach('get token to exit with', async () => {
+          tokenIndex = pool.bptIndex == 0 ? 1 : 0;
+          token = tokens.get(tokenIndex);
+        });
+
+        context('not in recovery mode', () => {
+          itExitsExactBptInForOneTokenOutProperly();
+        });
+
+        context('in recovery mode', () => {
+          sharedBeforeEach('enable recovery mode', async () => {
+            await pool.enableRecoveryMode(admin);
+          });
+
+          itExitsExactBptInForOneTokenOutProperly();
+        });
+
+        function itExitsExactBptInForOneTokenOutProperly() {
+          it('grants one token for exact bpt', async () => {
+            // 20% of previous balance
+            const previousBptBalance = await pool.balanceOf(lp);
+            const bptIn = pct(previousBptBalance, 0.2);
+            const expectedTokenOut = await pool.estimateTokenOutGivenBptIn(token, bptIn);
+
+            const result = await pool.singleExitGivenIn({ from: lp, bptIn, token });
+
+            // Only token out should be the one transferred
+            expect(result.amountsOut[tokenIndex]).to.be.equalWithError(expectedTokenOut, 0.0001);
+            expect(result.amountsOut.filter((_, i) => i != tokenIndex)).to.be.zeros;
+
+            const bptAfter = await pool.balanceOf(lp);
+
+            // Current BPT balance should decrease
+            expect(previousBptBalance.sub(bptIn)).to.equal(bptAfter);
+          });
+
+          it('can tell token amount returned (for either exit or exitSwap)', async () => {
+            const bptIn = pct(await pool.balanceOf(lp), 0.2);
+            const expectedTokenOut = await pool.estimateTokenOutGivenBptIn(token, bptIn);
+
+            const result = await pool.querySingleExitGivenIn({ bptIn, token });
+
+            expect(result.bptIn).to.equal(bptIn);
+            expect(result.amountsOut.filter((_, i) => i != tokenIndex)).to.be.zeros;
+            expect(result.amountsOut[tokenIndex]).to.be.equalWithError(expectedTokenOut, 0.0001);
+
+            // Exit swap should give the same number of tokens
+            const { amountOut } = await pool.swapGivenIn({
+              from: lp,
+              in: pool.bpt,
+              out: token,
+              amount: bptIn,
+              recipient: lp,
+            });
+            expect(amountOut).to.be.equalWithError(expectedTokenOut, 0.00001);
```

Comment:
This is an interesting test, but a) it should be its own test, not bundled with the other one, and b) we would ideally compare the actual swap with the actual join, not with the expected amounts. This cannot be done as the swap/join will alter the state, but we _can_ do it via queries (i.e. check that the swap query exactly equals the join query).

Note that these tests will fail for now, as the swap fees are not handled the same way in joins and swaps, but they should eventually pass.

Reply:
We should also test protocol fees here, but those cannot be done via queries. Maybe we can just create an issue for now and tackle that later on.

Reply:
Created the issue and updated the tests.